src/gui/symbology/qgscategorizedsymbolrendererwidget.cpp

```diff
@@ -1093,6 +1067,62 @@ void QgsCategorizedSymbolRendererWidget::deleteAllCategories()
   emit widgetChanged();
 }
 
+void QgsCategorizedSymbolRendererWidget::deleteUnusedCategories()
+{
+  if ( !mRenderer )
+    return;
+  const QString attrName = mExpressionWidget->currentField();
+  const QList<QVariant> uniqueValues = layerUniqueValues( attrName );
+
+  const QgsCategoryList catList = mRenderer->categories();
+
+  QList<int> unusedIndexes;
+
+  for ( int i = 0; i < catList.size(); ++i )
+  {
+    const QgsRendererCategory cat = catList.at( i );
+    if ( !uniqueValues.contains( cat.value() ) )
+    {
+      unusedIndexes.append( i );
+    }
+  }
+  mModel->deleteRows( unusedIndexes );
+  emit widgetChanged();
+}
+
+QList<QVariant> QgsCategorizedSymbolRendererWidget::layerUniqueValues( const QString &attrName )
+{
+  const int idx = mLayer->fields().lookupField( attrName );
+  QList<QVariant> uniqueValues;
+  if ( idx == -1 )
+  {
+    // Lets assume it's an expression
+    QgsExpression *expression = new QgsExpression( attrName );
```

Comment:
You have à memory leak here, please use std::unique_ptr here or remove the ptr .

Reply:
Ok, I have no problem with these suggestions and I'm happy to make these changes, but please note that this is part of the block of code which I reused from the existing `addCategories()` method. I.e. this code is in the current master branch:

https://github.com/qgis/QGIS/blob/87898417f79066168a792c94f3b7d907cbf52129/src/gui/symbology/qgscategorizedsymbolrendererwidget.cpp#L908

Again, I'm happy to change it as part of this PR- is there anything else that you think should be modified here?

Reply:
Ideally, I would have spotted during my first review , but thank you for making the modification, one less memory leak.

Lgtm

Reply:
Absolutely! Thanks again for your help @troopa81.

Reply:
Hi @troopa81, even though this PR is closed I just wanted to leave this here as a follow up. Firstly, I appreciate the time you gave me here- I should have picked these things up myself! As you can probably tell I'm fairly inexperienced in C++, but your reviews here prompted me to think/learn a bit more about memory management, heap/stack allocation etc. and to be a bit more analytical/critical when modifying existing QGIS code which will help me to submit better PRs in the future. So I just wanted to say thank you.

Reply:
@benwirf You're welcome! I'm always happy to see new contributors. Keep up the good work